crates/matrix-sdk/src/room_directory_search.rs

```diff
@@ -0,0 +1,370 @@
+        RoomDescription {
+            room_id: RoomId::parse("!ol19s:bleecker.street").unwrap(),
+            name: Some("CHEESE".into()),
+            topic: Some("Tasty tasty cheese".into()),
+            alias: None,
+            avatar_url: Some("mxc://bleeker.street/CHEDDARandBRIE".into()),
+            join_rule: ruma::directory::PublicRoomJoinRule::Public,
+            is_world_readable: true,
+            joined_members: 37,
+        }
+    }
+
+    fn get_second_page_description() -> RoomDescription {
+        RoomDescription {
+            room_id: RoomId::parse("!ca18r:bleecker.street").unwrap(),
+            name: Some("PEAR".into()),
+            topic: Some("Tasty tasty pear".into()),
+            alias: RoomAliasId::parse("#murrays:pear.bar").ok(),
+            avatar_url: Some("mxc://bleeker.street/pear".into()),
+            join_rule: ruma::directory::PublicRoomJoinRule::Knock,
+            is_world_readable: false,
+            joined_members: 20,
+        }
+    }
+
+    async fn new_server_and_client() -> (MockServer, Client) {
+        let server = MockServer::start().await;
+        let client = logged_in_client(Some(server.uri())).await;
+        (server, client)
+    }
+
+    #[async_test]
+    async fn search_success() {
+        let (server, client) = new_server_and_client().await;
+
+        let mut room_directory_search = RoomDirectorySearch::new(client);
+        Mock::given(RoomDirectorySearchMatcher { next_token: None, filter_term: None, limit: 1 })
+            .respond_with(ResponseTemplate::new(200).set_body_json(&*test_json::PUBLIC_ROOMS))
+            .mount(&server)
+            .await;
+
+        room_directory_search.search(None, 1).await.unwrap();
+        let (results, _) = room_directory_search.results();
+        assert_eq!(results.len(), 1);
+        assert_eq!(results[0], get_first_page_description());
+        assert!(!room_directory_search.is_at_last_page);
+        assert_eq!(room_directory_search.loaded_pages(), 1);
+    }
+
+    #[async_test]
+    async fn search_success_paginated() {
+        let (server, client) = new_server_and_client().await;
+
+        let mut room_directory_search = RoomDirectorySearch::new(client);
+        Mock::given(RoomDirectorySearchMatcher { next_token: None, filter_term: None, limit: 1 })
+            .respond_with(ResponseTemplate::new(200).set_body_json(&*test_json::PUBLIC_ROOMS))
+            .mount(&server)
+            .await;
+
+        room_directory_search.search(None, 1).await.unwrap();
+
+        Mock::given(RoomDirectorySearchMatcher {
+            next_token: Some("p190q".into()),
+            filter_term: None,
+            limit: 1,
+        })
+        .respond_with(
+            ResponseTemplate::new(200).set_body_json(&*test_json::PUBLIC_ROOMS_FINAL_PAGE),
+        )
+        .mount(&server)
+        .await;
+
+        room_directory_search.next_page().await.unwrap();
+
+        let (results, _) = room_directory_search.results();
```

Comment:
I think it would be nice to move `results()` after `search()`, and then use the opportunity to have the `Stream` to test it works correctly.